### DISCUSSION TITLE
Question: How to just print stdout and stderr when executing commands

### DISCUSSION THREAD

**Discussion by markusdd** [gitpython-developers/GitPython#1241] [Q&A]
Hi,

I am using gitpython for a small helper tool to manage somewhat complicated submodule setups.

I want it to work alongside standard git, not agaionst or around it.

Hence I want users to see the output of the commands. as they know it from plain vanilla git

e.g. a "Everything up to date" after a push without changes.

From the docs I was not really able to figure out a consistent way to do it.

For commit, I was able to achieve it like this: `print(toprepo.git.commit(a=pa.add, m=msg, with_stdout=True))`

For push, the same does not work, because it ends up on stderr `print(toprepo.git.push(pa.remote, with_stdout=True))`:
Logger produces this:
`INFO:git.cmd:git push origin -> 0; stdout: '<OUTPUT_STREAM>'; stderr: 'Everything up-to-date'`

When using the built in remote.push() and remote.pull() methosds I cannot find a way to retrieve the output at all, as with_stdout=False is hardcoded without a way to override it via kwargs.

Can you give guidance on how to cover my use-case?

**Comment by Byron** [comment]
Using `repo.git.<command>` syntax essentially [calls `git`][git-cmd] and captures stdout and stderr, returning both usually depending on the configuration.

When dealing with remotes there are facilities in place which capture stderr to make progress messages easier to use. Have a look [at the p'progress' portion of the advanced usage docs][remote-progress] for some details.

If the script is mostly for calling the git command, a shell script might be easier to work with than GitPython which has its own peculiarities when [calling git][git-execute].

[remote-progress]: https://gitpython.readthedocs.io/en/stable/tutorial.html#advanced-repo-usage
[git-cmd]: https://gitpython.readthedocs.io/en/stable/tutorial.html#using-git-directly
[git-execute]: https://github.com/gitpython-developers/GitPython/blob/572bbb39bf36fecb502c9fdf251b760c92080e1e/git/cmd.py#L676:L778

**Comment by markusdd** [comment]
Yes we currently do have a shell script but due to some requirements we have decided to switch to python, so
I am essentially coming from this direction.
I appreciate the facilities GitPython provides, so using it was a deliberate choice.

Essentially I also want all the return containers for use in further processing, in the process I just want the user to see the direct output of the called git actions.

For the repo.git. commands I think I understand how to use stdout, but where does stderr end up?

and for the remot.push/pull I have read the section you suggested, but I stil no not quite get on how to just print what would've been printed by the git command (and still receive the FetchInfo objects that contain success info)

**Reply by Byron** [reply]
Stderr is always configured as pipe, so it would never show up in the parent terminal.

https://github.com/gitpython-developers/GitPython/blob/33346b25c3a4fb5ea37202d88d6a6c66379099c5/git/cmd.py#L825

This is why it is always consumed in code and parsed on the fly, feeding into progress handlers.

https://github.com/gitpython-developers/GitPython/blob/33346b25c3a4fb5ea37202d88d6a6c66379099c5/git/remote.py#L826
https://github.com/gitpython-developers/GitPython/blob/33346b25c3a4fb5ea37202d88d6a6c66379099c5/git/remote.py#L698

These can also see the original line if the correct method is overridden.

https://github.com/gitpython-developers/GitPython/blob/33346b25c3a4fb5ea37202d88d6a6c66379099c5/git/remote.py#L737

Currently there is no way to not pipe stderr and display it to the enclosing terminal. You would rather have to launch the git command yourself.
